### PR TITLE
[builder/openstack] adds option to discover provisioning network

### PR DESCRIPTION
**Pull request by andreykaipov**
Closes https://github.com/hashicorp/packer/issues/8278.

For the full motivation behind this feature, please see the description in the linked issue above. The gist of it is not having users worry about providing different network IDs every time when switching between tenants.

Lemme know what y'all think!

### CODE REVIEW

File: builder/openstack/networks_test.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	testNot(t, "192.168.0.0/25", "192.168.0.0/24")
+
+	testYes(t, "192.168.101.202/16", "192.168.202.101/16")
+	testNot(t, "192.168.101.202/24", "192.168.202.101/24")
+	testNot(t, "192.168.202.101/24", "192.168.101.202/24")
+
+	testYes(t, "0.0.0.0/0", "192.168.0.0/24")
+	testYes(t, "0.0.0.0/0", "0.0.0.0/1")
+	testNot(t, "192.168.0.0/24", "0.0.0.0/0")
+	testNot(t, "0.0.0.0/1", "0.0.0.0/0")
```
Comment (azr) [inline builder/openstack/networks_test.go]: Oh, nice tests 🙂

File: builder/openstack/networks_test.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+	testYes(t, "2001:db8:fefe:b00b::/32", "2001:db8:b00b:fefe::/32")
+	testNot(t, "2001:db8:fefe:b00b::/64", "2001:db8:b00b:fefe::/64")
+	testNot(t, "2001:db8:b00b:fefe::/64", "2001:db8:fefe:b00b::/64")
+
+	testYes(t, "::/0", "2001:db8::/64")
+	testYes(t, "::/0", "::/1")
+	testNot(t, "2001:db8::/64", "::/0")
+	testNot(t, "::/1", "::/0")
+}
```
Comment (azr) [inline builder/openstack/networks_test.go]: Oh, nice readable tests 🙂

File: builder/openstack/step_discover_network.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	networks := make([]servers.Network, len(s.Networks)+len(s.Ports))
+	i := 0
+	for ; i < len(s.Ports); i++ {
+		networks[i].Port = s.Ports[i]
+	}
+	for ; i < len(networks); i++ {
+		networks[i].UUID = s.Networks[i-len(s.Ports)]
+	}
```
Comment (azr) [inline builder/openstack/step_discover_network.go]: Nit: this part looks complicated and could be simpler:

Suggestion:
```go
	networks := []servers.Network{} // or make([]servers.Network, 0, len(s.Networks)+len(s.Ports)) if we really wanted to optimize memory here, but I don't think this is necessary given what's after line 38
	for _, port := range s.Ports {
		networks = append(networks, servers.Network{Port : port})
	}
	for _, uuid := range s.Networks {
		networks = append(networks, servers.Network{UUID : uuid})
	}
```

Comment (andreykaipov) [reply builder/openstack/step_discover_network.go]: Thank you. Updated it :-)

**Comment by andreykaipov** [comment]
Thank you both for the quick review and acceptance! 😄

**Pull request merged**